File: src/gestorAplicacion/host/Cliente.java
package gestorAplicacion.host;

import java.io.Serializable;
import java.util.*;
import gestorAplicacion.enrutadorHFC.Router;
import gestorAplicacion.enrutadorHFC.Servidor;
import gestorAplicacion.servicio.Factura;
import gestorAplicacion.servicio.Plano;
import gestorAplicacion.enrutadorHFC.Antena;

public class Cliente implements Serializable{

  //CLASE PRINCIPAL YA QUE CASI TODAS LAS FUNCIONALIDADES ESTÁN ORIENTADAS A LOS CLIENTES

  //ATRIBUTOS
  private String nombre;
  private final long ID;
  private Router modem;
  private ProveedorInternet proveedor;
  private ArrayList<Integer> plan;
  private String nombrePlan;
  private Factura factura;
  

  //CONSTRUCTORES
  public Cliente(String nombre, long id, Router modem, ProveedorInternet proveedor, ArrayList<Integer> plan, String nombrePlan, Factura factura) {
    this.ID= id;
    this.nombre=nombre;
    this.modem = modem;
    this.proveedor = proveedor;
    this.plan = plan;
    this.nombrePlan = nombrePlan;
    this.factura = factura;
    proveedor.getClientes().add(this);
  }

  public Cliente(String nombre,long id, Router modem, ProveedorInternet proveedor, ArrayList<Integer> plan, String nombrePlan) { 

    this.ID= id;
    this.nombre=nombre;
    this.modem = modem;
    this.proveedor = proveedor;
    this.plan = plan;
    this.nombrePlan = nombrePlan;
    proveedor.getClientes().add(this);
  }
  
  public Cliente(String nombre, long id,ProveedorInternet proveedor) {
    this.nombre = nombre;
    ID = id;
    this.proveedor = proveedor;
    proveedor.getClientes().add(this);
  }

  
  //METODOS

  //METODO INSTANCIA--FUNCIONALIDAD ADQUISICION DE PLAN---RETORNA UN ARREGLO CON CLIENTES QUE PERTENECEN A UN PLAN ESPECÍFICO
  public ArrayList<Cliente> buscarCliente(ArrayList<Cliente> clientes, String nombrePlan) {
    ArrayList<Cliente> clientesPorPlan = new ArrayList<Cliente>();
    for (Cliente cliente : clientes) {
      if (cliente.getNombrePlan()!=null){
        if (cliente.getNombrePlan().equals(nombrePlan)) {
          clientesPorPlan.add(cliente);
        }
      }
      }
      return clientesPorPlan;
  }

  //METODO ESTÁTICO---FUNCIONALIDAD REPORTE---ETORNA UN ARREGLO CON CLIENTES QUE PERTENECEN A UN MISMO PROVEEDOR, SEDE Y PLAN
  public static ArrayList<Cliente> buscarCliente1(String sede,String nombrePlan,ProveedorInternet proveedor){
    ArrayList<Cliente> clientesPlan = new ArrayList<Cliente>();
    for(Cliente cliente: proveedor.getClientes()){
      if (cliente.getNombrePlan()!=null){
        if ((cliente.getNombrePlan().equals(nombrePlan)) && (cliente.getModem().getSede().equals(sede))){
          clientesPlan.add(cliente);
        }
      }
    }
    return clientesPlan;
  }

  //METODO INSTANCIA---FUNCIONALIDAD ADQUISICION DE PLAN---REALIZA TODOS LOS CAMBIOS NECESARIOS PARA REGISTRAR A UN CLIENTE NUEVO, BUSCA UN SERVIDOR Y UNA ANTENA DEL PROVEEDOR INDICADO PARA ASOCIARLOS AL NUEVO ROUTER 
  public Factura configurarPlan(String plan, Cliente cliente, String sede, int coordenadaX, int coordenadaY) {

    Servidor servidorAsociado=null;
    for(Servidor servidor: Servidor.getServidoresTotales()){
      if(servidor.getProveedor().getNombre().equals(cliente.getProveedor().getNombre())){
        if(servidor.getSede().equals(sede)){
          servidorAsociado=servidor;
        }
      }
    }
    
    if (plan.equals("BASIC")) {
      cliente.setNombrePlan(plan);
      cliente.setPlan(cliente.getProveedor().getPlanes().getBASIC());
      Router routercliente = new Router(cliente.getProveedor().getPlanes().getBASIC().get(0), cliente.getProveedor().getPlanes().getBASIC().get(1),true,servidorAsociado);
      cliente.setModem(routercliente);
      routercliente.setSede(sede);
      Plano planoCliente = new Plano(coordenadaX,coordenadaY);
      planoCliente.setSede(servidorAsociado.getCoordenadas().getSede());
      routercliente.setCoordenadas(planoCliente);

      for(Antena antena: Antena.getAntenasTotales()){
        if(antena.getSede().equals(sede)){
          if(antena.getProveedor().getNombre().equals(cliente.getProveedor().getNombre())){
            routercliente.setAntenaAsociada(antena);
            break;
          }
        }
      }
     
      
    } else if (plan.equals("STANDARD"))  {
      cliente.setNombrePlan(plan);
      cliente.setPlan(cliente.getProveedor().getPlanes().getSTANDARD());
      Router routercliente = new Router(cliente.getProveedor().getPlanes().getSTANDARD().get(0), cliente.getProveedor().getPlanes().getSTANDARD().get(1),true,servidorAsociado);
      cliente.setModem(routercliente);
      routercliente.setSede(sede);
      Plano planoCliente = new Plano(coordenadaX,coordenadaY);
      planoCliente.setSede(servidorAsociado.getCoordenadas().getSede());
      routercliente.setCoordenadas(planoCliente);

      for(Antena antena: Antena.getAntenasTotales()){
        if(antena.getSede().equals(sede)){
          if(antena.getProveedor().getNombre().equals(cliente.getProveedor().getNombre())){
            routercliente.setAntenaAsociada(antena);
            break;
          }
        }
      }
      
      
    }else{
      cliente.setNombrePlan(plan);
      cliente.setPlan(cliente.getProveedor().getPlanes().getPREMIUM());
      Router routercliente = new Router(cliente.getProveedor().getPlanes().getPREMIUM().get(0),cliente.getProveedor().getPlanes().getPREMIUM().get(1),true,servidorAsociado);
      cliente.setModem(routercliente);
      routercliente.setSede(sede);
      Plano planoCliente = new Plano(coordenadaX,coordenadaY);
      planoCliente.setSede(servidorAsociado.getCoordenadas().getSede());
      routercliente.setCoordenadas(planoCliente);
      
      for(Antena antena: Antena.getAntenasTotales()){
        if(antena.getSede().equals(sede)){
          if(antena.getProveedor().getNombre().equals(cliente.getProveedor().getNombre())){
            routercliente.setAntenaAsociada(antena);
            break;
          }
        }
      }
  
    }

    return new Factura(cliente);
  }

  //SETTERS Y GETTERS
  
  public String getNombre() {
    return nombre;
  }

  public void setNombre(String nombre) {
    this.nombre = nombre;
  }

  public long getID() {
    return ID;
  }

  public Router getModem() {
    return modem;
  }

  public void setModem(Router modem) {
    this.modem = modem;
  }

  public ProveedorInternet getProveedor() {
    return proveedor;
  }

  public void setProveedor(ProveedorInternet proveedor) {
    this.proveedor = proveedor;
  }

  public ArrayList<Integer> getPlan() {
    return plan;
  }

  public void setPlan(ArrayList<Integer> plan) {
    this.plan = plan;
  }

  public String getNombrePlan() {
    return nombrePlan;
  }

  public void setNombrePlan(String nombrePlan) {
    this.nombrePlan = nombrePlan;
  }

  public Factura getFactura() {
    return factura;
  }

  public void setFactura(Factura factura) {
    this.factura = factura;
  }

}

File: Python/src/gestorAplicacion/host/cliente.py
import sys

sys.path.append("src/gestorAplicacion/enrutadorHFC")
sys.path.append("src/gestorAplicacion/servicio")

from antena import Antena
from servidor import Servidor 
from router import Router
from antena import Antena
from factura import Factura
from plano import Plano

class Cliente:
  
  # //CLASE PRINCIPAL YA QUE CASI TODAS LAS FUNCIONALIDADES ESTÁN ORIENTADAS A LOS CLIENTES

  # CONSTRUCTOR

  def __init__(self, nombre=None, ID=0, proveedor=None, modem=None,  plan=[], nombrePlan=None, factura=None):
    
    # ATRIBUTOS
    self._nombre=nombre
    self._ID = ID
    self._modem = modem
    self._proveedor = proveedor
    self._plan = plan
    self._nombrePlan = nombrePlan
    self._factura = factura

    if proveedor!=None: #NO AÑADE CUANDO EL PROVEEDOR ES POR DEFECTO
      proveedor.getClientes().append(self)


  #METODOS

  # //METODO INSTANCIA--FUNCIONALIDAD ADQUISICION DE PLAN---RETORNA UNA LISTA CON CLIENTES QUE PERTENECEN A UN PLAN ESPECÍFICO
  def buscarCliente(self, clientes, nombrePlan):
    clientesPorPlan=[]
    for cliente in clientes:
      if cliente.getNombrePlan()!=None:
        if cliente.getNombrePlan() == nombrePlan:
          clientesPorPlan.append(cliente)
    return clientesPorPlan

  #  //METODO ESTÁTICO---FUNCIONALIDAD REPORTE---RETORNA UNA LISTA  CON CLIENTES QUE PERTENECEN A UN MISMO PROVEEDOR, SEDE Y PLAN
  @classmethod
  def buscarCliente1(cls,sede, nombrePlan, proveedor):
    clientesPlan=[]
    for cliente in proveedor.getClientes():
      if cliente.getNombrePlan()!=None:
        if (cliente.getNombrePlan() == nombrePlan) and (cliente.getModem().getSede() == sede):
          clientesPlan.append(cliente)

    return clientesPlan

  # //METODO INSTANCIA---FUNCIONALIDAD ADQUISICION DE PLAN---REALIZA TODOS LOS CAMBIOS NECESARIOS PARA REGISTRAR A UN CLIENTE NUEVO, BUSCA UN SERVIDOR Y UNA ANTENA DEL PROVEEDOR INDICADO PARA ASOCIARLOS AL NUEVO ROUTER 
  def configurarPlan(self, plan, cliente, proveedor, sede, coordenadaX, coordenadaY):

    servidorAsociado=None

    for servidor in Servidor.getServidoresTotales():
      if servidor.getProveedor().getNombre()==proveedor.getNombre():
        if servidor.getSede()==sede:
          servidorAsociado=servidor

    if plan=="BASIC":
      cliente.setNombrePlan(plan)
      cliente.setPlan(proveedor.getPlanes().getBASIC())
      routercliente = Router(proveedor.getPlanes().getBASIC()[0], proveedor.getPlanes().getBASIC()[1],True,servidorAsociado)
      cliente.setModem(routercliente)
      routercliente.setSede(sede)
      planoCliente = Plano(coordenadaX,coordenadaY)
      planoCliente.setSede(servidorAsociado.getCoordenadas().getSede())
      routercliente.setCoordenadas(planoCliente)

      for antena in Antena.getAntenasTotales():
        if antena.getSede()==sede:
          if antena.getProveedor().getNombre()==proveedor.getNombre():
            routercliente.setAntenaAsociada(antena)
            break  
            
    elif plan=="STANDARD":
      cliente.setNombrePlan(plan)
      cliente.setPlan(proveedor.getPlanes().getSTANDARD())
      routercliente = Router(proveedor.getPlanes().getSTANDARD()[0], proveedor.getPlanes().getSTANDARD()[1],True,servidorAsociado);
      cliente.setModem(routercliente)
      routercliente.setSede(sede)
      planoCliente = Plano(coordenadaX,coordenadaY)
      planoCliente.setSede(servidorAsociado.getCoordenadas().getSede())
      routercliente.setCoordenadas(planoCliente)

      for antena in Antena.getAntenasTotales():
        if antena.getSede()==sede:
          if antena.getProveedor().getNombre()==proveedor.getNombre():
            routercliente.setAntenaAsociada(antena)
            break  
    else:
      cliente.setNombrePlan(plan)
      cliente.setPlan(proveedor.getPlanes().getPREMIUM())
      routercliente = Router(proveedor.getPlanes().getPREMIUM()[0],proveedor.getPlanes().getPREMIUM()[1],True,servidorAsociado);
      cliente.setModem(routercliente)
      routercliente.setSede(sede)
      planoCliente = Plano(coordenadaX,coordenadaY)
      planoCliente.setSede(servidorAsociado.getCoordenadas().getSede())
      routercliente.setCoordenadas(planoCliente)

      for antena in Antena.getAntenasTotales():
        if antena.getSede()==sede:
          if antena.getProveedor().getNombre()==proveedor.getNombre():
            routercliente.setAntenaAsociada(antena)
            break
          
    return Factura(cliente)

  #SETTERS Y GETTERS
  def getNombre(self):
    return self._nombre
  
  def setNombre(self, nombre):
    self._nombre=nombre
    
  def getID(self):
    return self._ID

  def getModem(self):
    return self._modem
  
  def setModem(self, modem):
    self._modem=modem

  def getProveedor(self):
    return self._proveedor
  
  def setProveedor(self, proveedor):
    self._proveedor=proveedor

  def getPlan(self):
    return self._plan
  
  def setPlan(self,plan):
    self._plan=plan

  def getNombrePlan(self):
    return self._nombrePlan
  
  def setNombrePlan(self,nombrePlan):
    self._nombrePlan=nombrePlan

  def getFactura(self):
    return self._factura
  
  def setFactura(self,factura):
    self._factura=factura